### ISSUE TITLE
[Openshift] Fix Deployment

### ISSUE THREAD

**Issue by alexanderniebuhr** [democratic-csi/democratic-csi#36]
I am installing the driver on OKD based on Openshift, but it does not deploy completely because it uses HostPath.   
I am very new to CSI, sadly I am not very safe to investigate the code myself, but am happy to help and learn

```
Error creating: pods "zfs-iscsi-democratic-csi-node-" is forbidden: unable to validate against any security context constraint [
provider restricted: .spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used
provider restricted: .spec.securityContext.hostIPC: Invalid value: true: Host IPC is not allowed to be used 
spec.volumes[0]: Invalid value: "hostPath": hostPath volumes are not allowed to be used
spec.volumes[1]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[2]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[3]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[4]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[5]: Invalid value: "hostPath": hostPath volumes are not allowed to be used
spec.volumes[6]: Invalid value: "hostPath": hostPath volumes are not allowed to be used
spec.volumes[7]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[8]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[9]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[10]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.volumes[11]: Invalid value: "hostPath": hostPath volumes are not allowed to be used 
spec.containers[0].securityContext.privileged: Invalid value: true: Privileged containers are not allowed capabilities.add: Invalid value: "SYS_ADMIN": capability may not be added
spec.containers[0].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used 
spec.containers[0].securityContext.hostIPC: Invalid value: true: Host IPC is not allowed to be used
spec.containers[1].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used 
spec.containers[1].securityContext.hostIPC: Invalid value: true: Host IPC is not allowed to be used
spec.containers[2].securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used
spec.containers[2].securityContext.hostIPC: Invalid value: true: Host IPC is not allowed to be used]
```

**Comment by travisghansen** [comment]
I can look around but I’m not entirely sure how okd normally deals with this kind of thing. It requires very elevated permissions because it’s dealing with host level mounts and host level block devices doing formatting etc.

I’ll do some research but if you know generally what needs to be done let me know!

**Comment by alexanderniebuhr** [comment]
If HostPath is unavoidable then the containers in OKD / Openshift must run as privileged (see following)
However, the main reason to use OKD / Openshift is that container do not get elevated permissions. So another solution might be more future proof

- https://docs.okd.io/latest/storage/persistent_storage/persistent-storage-hostpath.html#persistent-storage-hostpath-pod_persistent-storage-hostpath
- https://docs.okd.io/latest/storage/persistent_storage/persistent-storage-hostpath.html#persistent-storage-hostpath-pod_persistent-storage-hostpath

**Comment by travisghansen** [comment]
Yeah it’s pretty much a requirement for csi drivers. I’m sure we’re simply missing some standard setup...I’m going to see how other csi drivers deploy (ie: trident from netapp etc) and if I can find what’s necessary to make it work.

**Comment by alexanderniebuhr** [comment]
Also `hostIPC` and `hostNetwork` might be a bigger problem here, as these are not allowed for containers by default SCCs (security context contraints). I also think others do not use them. I know that the `ebs.csi.aws.com` does work.

**Comment by travisghansen** [comment]
Those both are explicitly required actually to interact properly with host nfs and host iscsi daemons.

I’ll probably need you to try out variations with psp and scc to see what works (or view how other csi drivers you’re using are deploying those). To help me do you know which of those 2 is throwing the error? Scc or psp?

Forgive the ignorance with openshift :(

**Comment by alexanderniebuhr** [comment]
I have some more information for this. OKD / Openshift works with different SCCs. This said we could bind these to the service accounts as following example of `ebs.csi.aws.com` shows.

```
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: ebs-privileged-role
rules:
  - verbs:
      - use
    apiGroups:
      - security.openshift.io
    resources:
      - securitycontextconstraints
    resourceNames:
      - privileged

```

```
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: ebs-controller-privileged-binding
subjects:
  - kind: ServiceAccount
    name: aws-ebs-csi-driver-controller-sa
    namespace: openshift-cluster-csi-drivers
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: ebs-privileged-role
```

```
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: ebs-node-privileged-binding
subjects:
  - kind: ServiceAccount
    name: aws-ebs-csi-driver-node-sa
    namespace: openshift-cluster-csi-drivers
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: ebs-privileged-role
```

Does this help? And should I try to add these manually and let you know if they are fixing the issue?

**Comment by travisghansen** [comment]
Yeah if you could test that would be fantastic! I don’t have a openshift cluster to work with or I would gladly do the testing.

Can you send over the associated scc resource from the above examples?

**Comment by alexanderniebuhr** [comment]
Yeah, I am fine doing the testing. If I were knowing more about csi in general I would have made a PR right away.
Above did fix the permission issue, but we are not there yet (we have new errors).

1. css ressource

```yaml
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
allowHostDirVolumePlugin: false
allowHostIPC: false
allowHostNetwork: false
allowHostPID: false
allowHostPorts: false
allowPrivilegeEscalation: true
allowPrivilegedContainer: false
allowedCapabilities: null
apiVersion: security.openshift.io/v1
defaultAddCapabilities: null
fsGroup:
  type: MustRunAs
groups:
- system:authenticated
kind: SecurityContextConstraints
metadata:
  annotations:
    include.release.openshift.io/self-managed-high-availability: "true"
    kubernetes.io/description: restricted denies access to all host features and requires pods to be run with a UID, and SELinux context that are allocated to the namespace.  This is the most restrictive SCC and it is used by default for authenticated users.
    release.openshift.io/create-only: "true"
  creationTimestamp: "2020-12-14T09:47:06Z"
  generation: 1
  name: restricted
  resourceVersion: "1537"
  selfLink: /apis/security.openshift.io/v1/securitycontextconstraints/restricted
  uid: 4c7b1e57-80d5-45eb-8563-c490af7c2aeb
priority: null
readOnlyRootFilesystem: false
requiredDropCapabilities:
- KILL
- MKNOD
- SETUID
- SETGID
runAsUser:
  type: MustRunAsRange
seLinuxContext:
  type: MustRunAs
supplementalGroups:
  type: RunAsAny
users: []
volumes:
- configMap
- downwardAPI
- emptyDir
- persistentVolumeClaim
- projected
- secret
```

2. the new error

```
Error: container create failed:
time="2020-12-20T17:18:31Z" level=warning msg="exit status 1"
time="2020-12-20T17:18:31Z" level=error msg="container_linux.go:349: starting container process caused \"process_linux.go:449: container init caused \\\"rootfs_linux.go:58: mounting \\\\\\\"/etc/localtime\\\\\\\" to rootfs \\\\\\\"/var/lib/containers/storage/overlay/411107db5a425f6beb61d9bb669a667de3736055d6645afc3f7ca1d90769733a/merged\\\\\\\" at \\\\\\\"/var/lib/containers/storage/overlay/411107db5a425f6beb61d9bb669a667de3736055d6645afc3f7ca1d90769733a/merged/usr/share/zoneinfo/UCT\\\\\\\" caused \\\\\\\"not a directory\\\\\\\"\\\"\"" container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"rootfs_linux.go:58: mounting \\\"/etc/localtime\\\" to rootfs \\\"/var/lib/containers/storage/overlay/411107db5a425f6beb61d9bb669a667de3736055d6645afc3f7ca1d90769733a/merged\\\" at \\\"/var/lib/containers/storage/overlay/411107db5a425f6beb61d9bb669a667de3736055d6645afc3f7ca1d90769733a/merged/usr/share/zoneinfo/UCT\\\" caused \\\"not a directory\\\"\""
```

**Comment by travisghansen** [comment]
Ah that can be disabled I think..lemme check real quick what was done to make that optional...

**Comment by travisghansen** [comment]
Here: https://github.com/democratic-csi/charts/blob/master/stable/democratic-csi/values.yaml#L120

set it to null or false should do the trick..

**Comment by alexanderniebuhr** [comment]
Sure. So the question is, could you implement the role and rolebindings conditionally in the helm chart. So it is just installed if on OKD / OpenShift and not breaking Kubernetes?

**Comment by travisghansen** [comment]
Yeah of course! From what I can tell the only difference is the addition to the cluster role and the scc itself?

Were you able to test an actual pvc now?

**Comment by alexanderniebuhr** [comment]
Yes. pvc with iscsi seems to work fine. It does state `bound` in okd and in truenas it generated iscsi targets and volumes. I don't have anything to test snapshots and expansion on right now. But I do not think it might be a problem. So yeah it looks like adding the cluster role and role binding is the difference. **NOTE the privileged scc is default with any OKD / OpenShift deployment, so no need to add it.**

**Comment by travisghansen** [comment]
Ah I see regarding the scc. So all you did was add the additional stuff to the clusterrole right?

Did you actually have a pod attach to the pvc? I want to make sure the node operations are functioning.

**Comment by alexanderniebuhr** [comment]
I did actually added them, might be wrong. But was easier to test for me. I used following steps

1. add role

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: csi-privileged-role
rules:
  - verbs:
      - use
    apiGroups:
      - security.openshift.io
    resources:
      - securitycontextconstraints
    resourceNames:
      - privileged
```

2. add controller role binding

```yaml
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: csi-controller-privileged-binding
subjects:
  - kind: ServiceAccount
    name: zfs-iscsi-democratic-csi-controller-sa
    namespace: democratic-csi
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: csi-privileged-role
```

3. add node role binding

```yaml
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: csi-node-privileged-binding
subjects:
  - kind: ServiceAccount
    name: zfs-iscsi-democratic-csi-node-sa
    namespace: democratic-csi
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: csi-privileged-role
```

4. changed localtimeHostPath to null (this might also be necessary as all OKD / OpenShift deployments are based on coreOS, or might add alternate path in readme)

**Comment by travisghansen** [comment]
I’ll just add the additional stuff to already created clusterrole so it should ‘just work’.

Did you actually attach from a pod? Once I know that fully worked I’ll adjust the chart and push a new version for you to test (it should only take a couple seconds honestly).

**Comment by alexanderniebuhr** [comment]
I just tried. Did not work. But this is an issue with the nodes. I have to find out how to start iscsi service on okd provisioned nodes

**Comment by travisghansen** [comment]
OK cool. I'll commit an update to the chart and have you deploy with the chart and we'll go from there.

**Comment by travisghansen** [comment]
Ok, make sure you remove any of the manual stuff that you created and try again with version `0.6.1` of the chart.

https://github.com/democratic-csi/charts/commit/bfdf4b174b76c191f8057fba40394122ac3ed72a

```
--set node.rbac.openshift.privileged=true
```

^ should do the trick.

**Comment by alexanderniebuhr** [comment]
Yes. Installation works. pvc too. Mounting currently not. But I do not think this is an error with the install..

**Comment by travisghansen** [comment]
@alexanderniebuhr awesome! We'll leave this open until we figure out the mounting situation just to be sure. If you're running iscsi you do need to make sure the iscsi tools are installed on the host and the daemon(s) running as part of node prep (outside the scope of the driver). If you want to use multipath make sure that's installed/configured/running as desired as well.

Let me know how I can help.

**Comment by alexanderniebuhr** [comment]
yeah. I did checked this. I enabled them in my cluster machine config pool. So they are working and throwing me some errors. Might be a setup issue with my truenas. I did installed it just today and am not very far with configuration .. but it should work. However here are some errors i gathered. Not quite right in this issue though

```
Dec 20 19:03:26 eleron iscsid[824]: iscsid: conn 0 login rejected: initiator error - target not found (02/03)
Dec 20 19:03:26 eleron iscsid[824]: iscsid: Connection9:0 to [target: iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260] through [iface: default] is shutdown.
```

**Comment by travisghansen** [comment]
Can you send the full helm values you’re using (cleanse them of course)?

When I sit at my desk I’ll send a couple commands that can help us figure out what’s going on.

**Comment by alexanderniebuhr** [comment]
Here is my full helm values yaml. I think it might be good, but I am missing something. Also, it seems I cannot get any iscsi connection to truenas right now, so I must check this

```yaml
csiDriver:
  # should be globally unique for a given cluster
  name: "iscsi.csi.****.io"

# add note here about volume expansion requirements
storageClasses:
  - name: truenas-iscsi-csi
    defaultClass: false
    reclaimPolicy: Delete
    volumeBindingMode: WaitForFirstConsumer
    allowVolumeExpansion: true
    parameters:
      # for block-based storage can be ext3, ext4, xfs
      # for nfs should be nfs
      fsType: ext4

      # if true, volumes created from other snapshots will be
      # zfs send/received instead of zfs cloned
      # detachedVolumesFromSnapshots: "false"

      # if true, volumes created from other volumes will be
      # zfs send/received instead of zfs cloned
      # detachedVolumesFromVolumes: "false"

    mountOptions: []
    secrets:
      provisioner-secret:
      controller-publish-secret:
      node-stage-secret:
      node-publish-secret:
      controller-expand-secret:

# if your cluster supports snapshots you may enable below
volumeSnapshotClasses: []
#- name: freenas-iscsi-csi
#  parameters:
#  # if true, snapshots will be created with zfs send/receive
#  # detachedSnapshots: "false"
#  secrets:
#    snapshotter-secret:

node:
  driver:
    localtimeHostPath: null

driver:
  config:
    # please see the most up-to-date example of the corresponding config here:
    # https://github.com/democratic-csi/democratic-csi/tree/master/examples
    # YOU MUST COPY THE DATA HERE INLINE!
    driver: truenas-iscsi
    instance_id:
    httpConnection:
      protocol: http
      host: 172.20.62.51
      port: 80
      # use only 1 of apiKey or username/password
      # if both are present, apiKey is preferred
      # apiKey is only available starting in TrueNAS-12
      apiKey: "****"
      #username: root
      #password:
      allowInsecure: true
      # use apiVersion 2 for TrueNAS-12 and up (will work on 11.x in some scenarios as well)
      # leave unset for auto-detection
      apiVersion: 2
    sshConnection:
      host: 172.20.62.51
      port: 22
      username: root
      # use either password or key
      password: "****"
      #privateKey: |
      #  -----BEGIN RSA PRIVATE KEY-----
      #  ...
      #  -----END RSA PRIVATE KEY-----
    zfs:
      # can be used to override defaults if necessary
      # the example below is useful for TrueNAS 12
      #cli:
      #  sudoEnabled: true
      #
      #  leave paths unset for auto-detection
      #  paths:
      #    zfs: /usr/local/sbin/zfs
      #    zpool: /usr/local/sbin/zpool
      #    sudo: /usr/local/bin/sudo
      #    chroot: /usr/sbin/chroot

      # can be used to set arbitrary values on the dataset/zvol
      # can use handlebars templates with the parameters from the storage class/CO
      #datasetProperties:
      #  "org.freenas:description": "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}/{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
      #  "org.freenas:test": "{{ parameters.foo }}"
      #  "org.freenas:test2": "some value"

      # total volume name (zvol/<datasetParentName>/<pvc name>) length cannot exceed 63 chars
      # https://www.ixsystems.com/documentation/freenas/11.2-U5/storage.html#zfs-zvol-config-opts-tab
      # standard volume naming overhead is 46 chars
      # datasetParentName should therefore be 17 chars or less
      datasetParentName: default/okd/vols
      # do NOT make datasetParentName and detachedSnapshotsDatasetParentName overlap
      # they may be siblings, but neither should be nested in the other
      detachedSnapshotsDatasetParentName: default/okd/snaps
      # "" (inherit), lz4, gzip-9, etc
      zvolCompression: ""
      # "" (inherit), on, off, verify
      zvolDedup: ""
      zvolEnableReservation: false
      # 512, 1K, 2K, 4K, 8K, 16K, 64K, 128K default is 16K
      zvolBlocksize:
    iscsi:
      targetPortal: "172.20.62.51:3260"
      targetPortals: []
      # leave empty to omit usage of -I with iscsiadm
      interface:

      # MUST ensure uniqueness
      # full iqn limit is 223 bytes, plan accordingly
      # default is "{{ name }}"
      #nameTemplate: "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}-{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
      namePrefix: csi-
      nameSuffix: "-okd"
      # add as many as needed
      targetGroups:
        # get the correct ID from the "portal" section in the UI
        - targetGroupPortalGroup: 1
          # get the correct ID from the "initiators" section in the UI
          targetGroupInitiatorGroup: 1
          # None, CHAP, or CHAP Mutual
          targetGroupAuthType: None
          # get the correct ID from the "Authorized Access" section of the UI
          # only required if using Chap
          targetGroupAuthGroup:

      extentInsecureTpc: true
      extentXenCompat: false
      extentDisablePhysicalBlocksize: true
      # 512, 1024, 2048, or 4096,
      extentBlocksize: 512
      # "" (let FreeNAS decide, currently defaults to SSD), Unknown, SSD, 5400, 7200, 10000, 15000
      extentRpm: "SSD"
      # 0-100 (0 == ignore)
      extentAvailThreshold: 0
```

**Comment by travisghansen** [comment]
Ok looks sane enough. You can comment out the apiVersion as that is auto detected now (or should be).

Did you make sure the portal group and initiator group were created beforehand? Are you sure the service is actually enabled and running on truenas?

**Comment by alexanderniebuhr** [comment]
yes, i did that. does IP 0.0.0.0 not work?

**Comment by travisghansen** [comment]
Should be fine for the portal. Did you check allow all initiators for the initiator group?

I’m also wondering if you this this: https://jira.ixsystems.com/plugins/servlet/mobile#issue/NAS-108521

Can you download the script here: https://github.com/democratic-csi/democratic-csi/blob/master/contrib/ctld-config-watchdog-db.sh

to your nas, edit the vars at the top as appropriate and run it?

**Comment by travisghansen** [comment]
Send the output of the first iteration if you do run that script (I run it in the background as a temporary ‘service’ until the bug is fixed in a released version of truenas...12.0-U2 should have the fix).

**Comment by alexanderniebuhr** [comment]
Yes. Allow initiator is set.

I get following (seems so wrong). Ah note I am running truenas scale ... not truenas core (my bad should have said that earlier)

```bash
./script.sh: line 13: ctladm: command not found
ctl target count: 0
./script.sh: line 17: ctladm: command not found
ctl extent count: 0
./script.sh: line 21: ctladm: command not found
ctl lun count: 0
DB target count: 1
DB extent count: 1
DB lun count: 1
regen ctld config
reload ctld service
```

**Comment by travisghansen** [comment]
Ooooooh! Yeah that script won’t work on scale :) I had a weird issue connecting with scale at first as well and then somewhere along the way it cleaned up.

`cat /etc/scst.conf`

and

`journalctl -xef —unit scst`

will be your friends here. Try them out and look for anything crazy.

**Comment by alexanderniebuhr** [comment]
I will have a look if I am back from work, I'll keep you updated.

**Comment by travisghansen** [comment]
K sleepy time for me anyway. Good luck!

**Comment by alexanderniebuhr** [comment]
I appreciate your time and help so much. Haven't seen this in opensource for a long time :).  I did some debugging, can you tell me why my cluster is not allowed (seen in second log)

```bash
HANDLER vdisk_blockio {
    DEVICE csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd {
        filename /dev/zvol/default/okd/vols/pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96
        blocksize 512
        read_only 0
        usn 00505627e451000
        naa_id 0x6589cfc000000ce79bf53d35c52e7080
        prod_id "iSCSI Disk"
        t10_vend_id TrueNAS
        t10_dev_id 00505627e451000
    }

}

TARGET_DRIVER iscsi {
    enabled 1

    TARGET iqn.2019-05.io.****.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd {
        enabled 1
        per_portal_acl 1

        GROUP security_group {

            LUN 0 csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd
        }
    }
}
```

```bash
Dec 20 20:03:26 spifly.nbhr.io iscsi-scstd[124623]: Connect from 172.20.62.120:44758 to 172.20.62.51:3260
Dec 20 20:03:26 spifly.nbhr.io iscsi-scstd[124623]: Initiator iqn.1994-05.com.redhat:1668411269eb not allowed
to connect to target iqn.2019-05.io.****:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd
```

**Comment by travisghansen** [comment]
Interesting...are you tracking nightlies? A couple of things to check:

1. Make sure you indeed have created the initiator group and have simply selected the 'allow all initiators' checkbox
2. Open the target and make sure that initiator ID shows up properly as selected in the iscsi group

In my case, I have the following:

```
TARGET_DRIVER iscsi {
    enabled 1
    TARGET iqn.2005-10.org.freenas.ctl:csi-pvc-111-onegear {
        enabled 1
        per_portal_acl 1
        GROUP security_group {
            INITIATOR *\#*
            LUN 0 csi-pvc-111-onegear
        }
    }
}
```

Of note in your case is the missing `INITIATOR *\#*` line in the security group.

**Comment by alexanderniebuhr** [comment]
So i fixed the target group now I get following in journalctl

```bash
Dec 21 19:32:12 spifly.****.io iscsi-scstd[124623]: Connect from 172.20.62.120:50382 to 172.20.62.51:3260
Dec 21 19:32:13 spifly.****.io iscsi-scstd[124623]: Connect from 172.20.62.120:50394 to 172.20.62.51:3260
Dec 21 19:32:14 spifly.****.io iscsi-scstd[124623]: Connect from 172.20.62.120:50414 to 172.20.62.51:3260
Dec 21 19:32:16 spifly.****.io iscsi-scstd[124623]: Connect from 172.20.62.120:50438 to 172.20.62.51:3260
Dec 21 19:32:20 spifly.****.io iscsi-scstd[124623]: Connect from 172.20.62.120:50474 to 172.20.62.51:3260
Dec 21 19:32:28 spifly.****.io iscsi-scstd[124623]: Connect from 172.20.62.120:50556 to 172.20.62.51:3260
Dec 21 19:32:44 spifly.****.io iscsi-scstd[124623]: Connect from 172.20.62.120:50704 to 172.20.62.51:3260
```

but still getting error in pod

```bash
MountVolume.MountDevice failed for volume "pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96" : rpc error: code = Internal desc = {"code":19,"stdout":"Logging in to [iface: default, target: iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260]\n","stderr":"iscsiadm: Could not login to [iface: default, target: iqn.2019-05.io.****:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260].\niscsiadm: initiator reported error (19 - encountered non-retryable iSCSI login failure)\niscsiadm: Could not log into all portals\n"}
```

**Comment by travisghansen** [comment]
Can you email me screenshots of the relevant assets in the iscsi screen? This is pretty weird for sure :(

**Comment by alexanderniebuhr** [comment]
I can make some. One quick question, does the InitiatorName does need to match any other iqn name?

**Comment by alexanderniebuhr** [comment]
Should be fine here through. No private data...

<img width="976" alt="Bildschirmfoto 2020-12-21 um 19 49 41" src="https://user-images.githubusercontent.com/45965090/102811508-bb01ae80-43c5-11eb-9819-8c8909bcdf40.png">
<img width="1403" alt="Bildschirmfoto 2020-12-21 um 19 49 24" src="https://user-images.githubusercontent.com/45965090/102811513-bc32db80-43c5-11eb-9198-fe5e174f32ba.png">
<img width="974" alt="Bildschirmfoto 2020-12-21 um 19 49 07" src="https://user-images.githubusercontent.com/45965090/102811514-bccb7200-43c5-11eb-9b7a-0c1f8a09d32f.png">
<img width="1404" alt="Bildschirmfoto 2020-12-21 um 19 48 53" src="https://user-images.githubusercontent.com/45965090/102811515-bd640880-43c5-11eb-91d2-17a888f778b3.png">

**Comment by travisghansen** [comment]
Name shouldn’t matter no. Can you send the extents and associated targets as well?

**Comment by alexanderniebuhr** [comment]
I am also found out a way to get on the nodes itself, so I could run iscsiadm commands from OpenShift nodes if this does help

<img width="826" alt="Bildschirmfoto 2020-12-21 um 20 00 06" src="https://user-images.githubusercontent.com/45965090/102812400-37e15800-43c7-11eb-8ff2-d36adceb7e4d.png">
<img width="822" alt="Bildschirmfoto 2020-12-21 um 19 59 52" src="https://user-images.githubusercontent.com/45965090/102812402-39128500-43c7-11eb-82d0-9702625ca28d.png">

**Comment by travisghansen** [comment]
You can get the logs from the `node` pod by using something like the following:

```
kubectl -n democratic-csi logs -f -l app.kubernetes.io/csi-role=node,app.kubernetes.io/instance=zfs-iscsi --container=csi-driver --max-log-requests=500
```

Replace the namespace and the instance with whatever appropriate details you have for your helm deployment. If you watch/get the logs you'll see exactly the `iscsiadm` commands getting executed. See if you find anything suspicious there. Also, if you have another node outside the cluster running iscsi it would be good to attempt attaching from that to see if we can narrow down the problem (server side or client side).

**Comment by alexanderniebuhr** [comment]
Ok logs do not really show any new error.
I deployed new Ubuntu VM outside, and got same issue.. so this might be server issue :(
I am very **embarrassed**. Normally I have no problems with iSCSI

**Comment by travisghansen** [comment]
No worries! I had a similar weird issue with iscsi on scale. I reported it and started discussing it with ix (on their slack channel) and literally when I started revisiting the problem after I initially encountered it a week or 2 previous things were working as expected...no changes on my end. So there could be something crazy going on for sure.

I'll fire up my scale box again in the next day or so and get it updated etc and try again to make sure things are still working.

If you find out anything on your end let me know..

**Comment by alexanderniebuhr** [comment]
Ok well I did the `turn off & turn on` and I got an step further, however now it looks like an client side issue again.

From my Ubuntu VM outside the cluster this works:

```bash
alexanderniebuhr@test:~$ sudo iscsiadm --mode discovery
spifly.nbhr.io:3260 via sendtargets

alexanderniebuhr@test:~$ sudo iscsiadm --mode discovery --type sendtargets --portal 172.20.62.51
172.20.62.51:3260,1 iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd

alexanderniebuhr@test:~$ sudo iscsiadm -m node -T iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 --login
Logging in to [iface: default, target: iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260]
Login to [iface: default, target: iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260] successful.

alexanderniebuhr@test:~$ sudo iscsiadm -m node -T iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 --login
Logging in to [iface: default, target: iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260]
Login to [iface: default, target: iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260] successful.
```

Hower from the csi driver logs following commands do not work, even not from my ubuntu VM

```bash
alexanderniebuhr@test:~$ sudo iscsiadm -m node -T iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 -o new
New iSCSI node [tcp:[hw=,ip=,net_if=,iscsi_if=default] 172.20.62.51,3260,-1 iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd] added

alexanderniebuhr@test:~$ sudo iscsiadm -m node -T iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 -o update --name node.startup --value manual

alexanderniebuhr@test:~$ sudo iscsiadm -m node -T iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 -l
Logging in to [iface: default, target: iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260]
iscsiadm: Could not login to [iface: default, target: iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd, portal: 172.20.62.51,3260].
iscsiadm: initiator reported error (19 - encountered non-retryable iSCSI login failure)
iscsiadm: Could not log into all portals
```

Do you have any idea?

**Comment by travisghansen** [comment]
If I understand correctly, the ubuntu VM works fine with using sendtargets/discovery, but fails when manually adding the node?

Can you try the sendtargets sequence directly on the openshift nodes to see if that works?

**Comment by alexanderniebuhr** [comment]
Not directly in the send targets. The difference between working and not working is in following connection line

1)
```bash
sudo iscsiadm -m node -T iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 --login
```

2)
```bash
sudo iscsiadm -m node -T iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 -o new
sudo iscsiadm -m node -T iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 -o update --name node.startup --value manual
sudo iscsiadm -m node -T iqn.2019-05.io.nbhr:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd -p 172.20.62.51:3260 -l
```

Number 1 works from external Ubuntu VM and OpenShift Node, number 2 does not work at all.

**Comment by alexanderniebuhr** [comment]
Note the difference in the target name (marked with stars)

```bash
iqn.2019-05.io.nbhr.spifly:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-okd
iqn.2019-05.io.nbhr******:storage:spifly:csi-pvc-0db8d5e3-23f7-4496-b62d-0d2fffc84d96-ok
```

**Comment by travisghansen** [comment]
Well that's problematic. Did you happen to change the name in the global configuration after creating the volume? I pull that info from the TrueNAS API when the volume is created.

Enable debug logging on the controller and then you'll see all the HTTP traffic etc back and forth between the controller and TrueNAS. Delete the PVC and recreate and send over the details that you can...we'll see what it responds with.

**Comment by alexanderniebuhr** [comment]
Seemed to be this issue, that I renamed it. :O 
Also seems that I have to restart service like 3 times before it works. However it does seem to claim and be able to attach to database pod.. 
This was very strange and I am so sorry with bothering you

**Comment by travisghansen** [comment]
Duuuuuuuude! So it worked?

The value cannot be changed after the volume is created as the data gets stored on the PV in k8s and stored on the zvol using zfs properties. In short, changing it in the TrueNAS UI will break already-created vols as it relates to democratic-csi.

**Comment by alexanderniebuhr** [comment]
Yes. I must do some advanced testing, but I think everything new will be worth a new issue.

I got it. Renaming isn't a problem, was just a troubleshooting step I made. So, because other issues where triggering I did make some later issues myself.

Thanks so much for your Time and help, I think we could close this now. However, I do think it might need some more documentation for openshift, as it is not "install and ready to go". However, as openshift and OKD have billion ways of Installation I don't think there is a way for all of them, so might be hard to do good usage documentation. The new helm option is great and might fix this for some of them because everything else seems to be my setup and truenas scale.

So, feel free to close 👍
